### DISCUSSION TITLE
Returning 500 YARP, returns 200 manually through Service Fabric reverse proxy

### DISCUSSION THREAD

**Discussion by WhitWaldo** [dotnet/yarp#302] [General] [answered]
### Describe the bug
I've set up a solution using YARP as a gateway reverse proxy between external requests and the reverse proxy intended to be utilized internally by Service Fabric. When running a 5-node cluster in local debug, I can issue a request to the project housing YARP and see the logging indicating that it's proxying the request to the appropriate endpoint, but it yields a 500. 

Sending a request via Postman to the same address YARP uses works with a 200. 

### To Reproduce
I've created a sample solution [SfYarp.zip](https://github.com/microsoft/reverse-proxy/files/4878197/SfYarp.zip) that demonstrates the issue. You'll need to run it on a machine with the Service Fabric SDK installed, but otherwise just hit F5 in Visual Studio to start it up. 

1) Initiate a request to `http://localhost:8603/weather` and you'll see something like the following in debug:

> Microsoft.AspNetCore.Hosting.Diagnostics: Information: Request starting HTTP/1.1 GET http://localhost:8603/weather  
> Microsoft.AspNetCore.Routing.EndpointMiddleware: Information: Executing endpoint 'forecast'
> Microsoft.ReverseProxy.Middleware.LoadBalancingMiddleware: Information: Operation started: ReverseProxy.PickDestination
> Microsoft.ReverseProxy.Middleware.LoadBalancingMiddleware: Information: Operation ended: ReverseProxy.PickDestination, 3.1505ms, success
> Microsoft.ReverseProxy.Middleware.ProxyInvokerMiddleware: Information: Operation started: ReverseProxy.Proxy
> Microsoft.ReverseProxy.Service.Proxy.HttpProxy: Information: Proxying to http://localhost:19081/SfYarp/OtherService/WeatherForecast
> Microsoft.ReverseProxy.Middleware.ProxyInvokerMiddleware: Information: Operation ended: ReverseProxy.Proxy, 115.7036ms, success
> Microsoft.AspNetCore.Routing.EndpointMiddleware: Information: Executed endpoint 'forecast'
> Microsoft.AspNetCore.Hosting.Diagnostics: Information: Request finished in 165.0948ms 500 

Note that line 6 contains the URL it's intended to proxy to: `http://localhost:19081/SfYarp/OtherService/WeatherForecast` This is as expected, but you'll note that it concludes with a 500.

2) Initiate a GET request manually to this URL and you'll get a 200 OK with a JSON response. 

When the request is performed manually, it works fine. When performed via YARP, it fails.

There are no exceptions observed anywhere, just the 500 internal service error when attempting to use YARP.

### Further technical details
Using .NET Core 3.1 on Windows 10 Pro using VS 16.6.3
Microsoft.ServiceFabric.AspNetCore.Kestrel - 4.1.417
Microsoft.ReverseProxy - 1.0.0-preview.2.20318.2

**Comment by Tratcher** [comment]
Are there any more hints of you turn the log level up to Debug?
What do the logs in the destination server say?

**Comment by WhitWaldo** [comment]
I'm not in terribly familiar territory here, so bear with me. Finally got some additional information to appear in the other service while debugging it and I see this error:

> Microsoft.ReverseProxy.Middleware.ProxyInvokerMiddleware: Information: Operation ended: ReverseProxy.Proxy, 162.1257ms, success
> Microsoft.AspNetCore.Routing.EndpointMiddleware: Information: Executed endpoint 'forecast'
> Microsoft.AspNetCore.Hosting.Diagnostics: Information: Request finished in 225.4117ms 500 
> 
> System.IO.IOException: Failed to bind to address http://[::]:8603: address already in use.
>  ---> Microsoft.AspNetCore.Connections.AddressInUseException: Only one usage of each socket address (protocol/network address/port) is normally permitted.
>  ---> System.Net.Sockets.SocketException (10048): Only one usage of each socket address (protocol/network address/port) is normally permitted.
>    at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, String callerName)
>    at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
>    at System.Net.Sockets.Socket.Bind(EndPoint localEP)
>    at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketConnectionListener.Bind()
>    --- End of inner exception stack trace ---
>    at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketConnectionListener.Bind()
>    at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransportFactory.BindAsync(EndPoint endpoint, CancellationToken cancellationToken)
>    at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer.<>c__DisplayClass21_0`1.<<StartAsync>g__OnBind|0>d.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(ListenOptions endpoint, AddressBindContext context)
>    --- End of inner exception stack trace ---
>    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(ListenOptions endpoint, AddressBindContext context)
>    at Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions.BindAsync(AddressBindContext context)
>    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.EndpointsStrategy.BindAsync(AddressBindContext context)
>    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindAsync(IServerAddressesFeature addresses, KestrelServerOptions serverOptions, ILogger logger, Func`2 createBinding)
>    at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServer.StartAsync[TContext](IHttpApplication`1 application, CancellationToken cancellationToken)

I recall noting in the [Service Fabric documentation](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-communication-aspnetcore#kestrel-in-reliable-services) that "_...unlike HTTP.sys, Kestrel doesn't support port sharing between multiple processes. Each instance of Kestrel must use a unique port._"

Is it possible that's the issue here? I'm going to try swapping out the use of Kestrel with HTTP.sys and see if that remedies this.

**Comment by WhitWaldo** [comment]
Yep, that did it - simply switching out the use of Kestrel with HTTP.sys did the charm. Here's a working variant of the original solution:
[SfYarp-httpsys.zip](https://github.com/microsoft/reverse-proxy/files/4880869/SfYarp-httpsys.zip)

Does this mean that Kestrel isn't a supported platform for YARP and that any implementations must use HTTP.sys instead?

**Comment by davidni** [comment]
There are multiple things at play, let's sort them out one at a time.

### 1. Port 8603 already in use
> `IOException: Failed to bind to address http://[::]:8603: address already in use`

@WhitWaldo you must have something else running that already claimed that port. I see that you are (correctly) setting Gateway's InstanceCount to 1 when running locally, so it is not immediately clear who the culprit is. Sugg running `netstat -abno` from an elevated prompt to identify who is LISTENING on `:8603` next time you see this error.

### 2. 500 status code

Since your localhost communication between Gateway and SF Reverse Proxy is unencrypted, you can easily capture it using a tool like Wireshark. Just compare that to a trace where you hit it with postman. you can also inspect SF Reverse Proxy diagnostics. see [docs](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reverse-proxy-diagnostics)

**Comment by Tratcher** [comment]
> Does this mean that Kestrel isn't a supported platform for YARP and that any implementations must use HTTP.sys instead?

Kestrel should work with YARP, that's what we use by default. Port conflicts are usually a configuration issue.

**Comment by WhitWaldo** [comment]
@davidni
netstat yields two listeners on 8603 when I launch the SF solution:

>   TCP    0.0.0.0:8603           0.0.0.0:0              LISTENING       92500
>  [Gateway.exe]
>   TCP    [::]:8603              [::]:0                 LISTENING       92500
>  [Gateway.exe]

From what I can find, this means that it's listening on all possible IPv4 addresses and the IPv6 loopback for port 8603. Does this count as multiple uses of the port?

I thought I'd see if I could just set the Kestrel options to use so it'd specifically just use 127.0.0.1 on 8603 instead of 0.0.0.0 or [::]
```
.UseKestrel(opt => {
  opt.Listen(IPAddress.Lookback, 8603);
});
```

But this just added another listener explicitly using 127.0.0.1 on top of the other two. Looking at the Wireshark traffic, it just led to the same binding error as posted before:

> 410	4.024712	127.0.0.1	127.0.0.1	HTTP	277	GET /weather HTTP/1.1 
> 411	4.024763	127.0.0.1	127.0.0.1	TCP	44	8603 → 61259 [ACK] Seq=1 Ack=235 Win=10232 Len=0
> 417	4.070687	::1	::1	TCP	76	61345 → 19081 [SYN] Seq=0 Win=65535 Len=0 MSS=65475 WS=256 SACK_PERM=1
> 418	4.070769	::1	::1	TCP	76	19081 → 61345 [SYN, ACK] Seq=0 Ack=1 Win=65535 Len=0 MSS=65475 WS=256 SACK_PERM=1
> 419	4.070817	::1	::1	TCP	64	61345 → 19081 [ACK] Seq=1 Ack=1 Win=2618880 Len=0
> 420	4.071542	::1	::1	TCP	441	61345 → 19081 [PSH, ACK] Seq=1 Ack=1 Win=2618880 Len=377 [TCP segment of a reassembled PDU]
> 421	4.071570	::1	::1	TCP	64	19081 → 61345 [ACK] Seq=1 Ack=378 Win=2618880 Len=0
> 422	4.071714	::1	::1	HTTP	69	GET /SfYarp/OtherService/WeatherForecast HTTP/1.1 
> 423	4.071729	::1	::1	TCP	64	19081 → 61345 [ACK] Seq=1 Ack=383 Win=2618880 Len=0
> 424	4.072506	::1	::1	HTTP	189	HTTP/1.1 500 Internal Server Error 
> 425	4.072548	::1	::1	TCP	64	61345 → 19081 [ACK] Seq=383 Ack=126 Win=2618624 Len=0
> 426	4.086716	127.0.0.1	127.0.0.1	HTTP	169	HTTP/1.1 500 Internal Server Error 

As a last attempt to get it to quit listening on multiple ports, I updated the `.UseUrls(url)` to use "http://localhost:8603" instead of the default being passed into it "http://+:8603" but netstat still indicates that it's listening on "127.0.01:8603" as well as "[::]:8603" and I'm not seeing how one might disable one or the other to reduce it to only a single listener. 

I'm not really seeing anything special you have in your samples regarding Kestrel configuration and I'm not really doing anything special here apart from running everything within Service Fabric. Any thoughts on why this isn't working within an SF environment? I'd wait on that SF integration PR, but I don't see anything in there that the Kestrel configuration to demonstrate what's not right here.

**Comment by Tratcher** [comment, marked as answer]
Kestrel can be configured via code, config file, or UseUrls. Is it possible you have the same url defined in both code and the config file?

**Comment by WhitWaldo** [comment]
Don't think so. The only sections in my appsettings.json (zipped two sample solutions in the above responses) are Logging, AllowedHosts and ReverseProxy. My WebHostBuilder is equivalent to what the Service Fabric template kicks out + logging:

```c#
        /// <summary>
        /// Optional override to create listeners (like tcp, http) for this service instance.
        /// </summary>
        /// <returns>The collection of listeners.</returns>
        protected override IEnumerable<ServiceInstanceListener> CreateServiceInstanceListeners()
        {
            return new ServiceInstanceListener[]
            {
                new ServiceInstanceListener(serviceContext =>
                    new KestrelCommunicationListener(serviceContext, "ServiceEndpoint", (url, listener) => 
                    //new HttpSysCommunicationListener(serviceContext, "ServiceEndpoint", (url, listener) =>
                    {
                        ServiceEventSource.Current.ServiceMessage(serviceContext, $"Starting Kestrel on {url}");
                        //ServiceEventSource.Current.ServiceMessage(serviceContext, $"Starting HTTP.sys on {url}");

                        return new WebHostBuilder()
                            //.UseHttpSys()
                            .UseKestrel()
                            .ConfigureServices(
                                services => services
                                    .AddSingleton<StatelessServiceContext>(serviceContext))
                            .ConfigureLogging(logging =>
                            {
                                logging.ClearProviders();
                                logging.AddDebug();
                            })
                            .UseContentRoot(Directory.GetCurrentDirectory())
                            .UseStartup<Startup>()
                            .UseServiceFabricIntegration(listener, ServiceFabricIntegrationOptions.None)
                            .UseUrls(url)
                            .Build();
                    }))
            };
        }
```

The input URL just gets set to `http://+:<port>` where the port is set in the SF ServiceManifest:

```
>   <Resources>
>     <Endpoints>
>       <!-- This endpoint is used by the communication listener to obtain the port on which to 
>            listen. Please note that if your service is partitioned, this port is shared with 
>            replicas of different partitions that are placed in your code. -->
>       <Endpoint Protocol="http" Name="ServiceEndpoint" Type="Input" Port="8603" UriScheme="http" />
>     </Endpoints>
>   </Resources>
```

And that's it - can't find anything else that deals with Kestrel configuration.
